sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/implementation/util/StorageBlockingSink.java

```diff
@@ -0,0 +1,96 @@
+package com.azure.storage.blob.implementation.util;
+
+import com.azure.core.util.logging.ClientLogger;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Sinks;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Queue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This is a lightweight wrapper around Sinks.Many to implement a Sink backed by a LinkedBlockingQueue, effectively
+ * implementing a BlockingSink.
+ */
+public class StorageBlockingSink {
+    private final ClientLogger logger = new ClientLogger(StorageBlockingSink.class);
+
+    /*
+    Note: Though this is used only by BlockBlobOutputStream, the decision was made to abstract this type out to make
+    testing easier.
+    */
+    private final Sinks.Many<ByteBuffer> writeSink;
+    private static final int WRITE_RETRY_IN_SECONDS = 3;
+
+    /**
+     * Create a new StorageBlockingSink.
+     */
+    public StorageBlockingSink() {
+        /*
+        We use a LinkedBlockingQueue to effectively block the thread that calls tryEmitNext if there is
+        backpressure from downstream. Its capacity is 1 to keep the buffer as small as possible, as downstream
+        implementations do their own buffering.
+        */
+        Queue<ByteBuffer> writeLimitQueue = new LinkedBlockingQueue<>(1);
+        this.writeSink = Sinks.many().unicast().onBackpressureBuffer(writeLimitQueue);
+    }
+
+    /**
+     * Try to emit an element.
+     *
+     * @param buffer {@link ByteBuffer} to emit.
+     * @return an optional IOException if an unrecoverable error was thrown.
+     */
+    public IOException tryEmitNext(ByteBuffer buffer) {
```

Comment:
I kind of dont like returning the Exception here but Im not sure how to make this better.

Reply:
public void tryEmitNext () throws IOException ?

Reply:
Yeah this is the best option

Reply:
resolved this